contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1051 @@
+        principalTotal = initialPrincipal;
+        weightForBucket[0] = weightForBucket[0].add(initialPrincipal);
+        weightForBucketForAccount[0][from] =
+            weightForBucketForAccount[0][from].add(initialPrincipal);
+
+        return address(this);
+    }
+
+    /**
+     * Called by Margin when additional value is added onto the position this contract
+     * is lending for. Balance is added to the address that loaned the additional tokens.
+     *
+     * @param  payer           Address that loaned the additional tokens
+     * @param  positionId      Unique ID of the position
+     * @param  principalAdded  Amount that was added to the position
+     *  param  lentAmount      (unused)
+     * @return                 This address to accept, a different address to ask that contract
+     */
+    function increaseLoanOnBehalfOf(
+        address payer,
+        bytes32 positionId,
+        uint256 principalAdded,
+        uint256 lentAmount
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        require(
+            payer == address(this),
+            "BucketLender#increaseLoanOnBehalfOf: Other lenders cannot lend for this position"
+        );
+        require(
+            !Margin(DYDX_MARGIN).isPositionCalled(POSITION_ID),
+            "BucketLender#increaseLoanOnBehalfOf: No lending while the position is margin-called"
+        );
+        require(
+            lentAmount <= availableTotal,
+            "BucketLender#increaseLoanOnBehalfOf: No lending not-accounted-for funds"
+        );
+
+        uint256 principalAfterIncrease = getCurrentPrincipalFromMargin();
+        uint256 principalBeforeIncrease = principalAfterIncrease.sub(principalAdded);
+
+        // principalTotal was the principal after the last increase
+        accountForClose(principalTotal.sub(principalBeforeIncrease));
+
+        accountForIncrease(principalAdded, lentAmount);
+
+        assert(principalTotal == principalAfterIncrease);
+
+        return address(this);
+    }
+
+    /**
+     * Function a contract must implement in order to let other addresses call marginCall().
+     *
+     * @param  caller         Address of the caller of the marginCall function
+     * @param  positionId     Unique ID of the position
+     * @param  depositAmount  Amount of heldToken deposit that will be required to cancel the call
+     * @return                This address to accept, a different address to ask that contract
+     */
+    function marginCallOnBehalfOf(
+        address caller,
+        bytes32 positionId,
+        uint256 depositAmount
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        require(
+            TRUSTED_MARGIN_CALLERS[caller],
+            "BucketLender#marginCallOnBehalfOf: Margin-caller must be trusted"
+        );
+        require(
+            depositAmount == 0,
```

Comment:
We should put in the comment of `Margin#marginCall` that making `depositAmount` 0 makes it so the position owner can't cancel it by depositing. It took some digging for me to confirm that

Reply:
done